lib/src/backend/mod.rs

```diff
@@ -0,0 +1,57 @@
+use anyhow::{anyhow, Result};
+use clap::ValueEnum;
+
+use ostree::glib;
+use ostree_ext::container::OstreeImageReference;
+use ostree_ext::keyfileext::KeyFileExt;
+use ostree_ext::ostree;
+use ostree_ext::sysroot::SysrootLock;
+
+use crate::spec::ImageStatus;
+
+mod ostree_container;
+
+pub(crate) trait Backend {
+    fn backend(&self) -> Result<Box<dyn BackendImpl>>;
+}
+
+pub(crate) trait BackendImpl {
+    fn imagestatus(
+        &self,
+        sysroot: &SysrootLock,
+        deployment: &ostree::Deployment,
+        image: OstreeImageReference,
+    ) -> Result<(Option<ImageStatus>, Option<ImageStatus>)>;
+}
+
+impl Backend for crate::spec::Backend {
+    fn backend<'a>(&self) -> Result<Box<dyn BackendImpl>> {
+        match self {
+            crate::spec::Backend::OstreeContainer => {
+                Ok(Box::new(ostree_container::OstreeContainerBackend))
+            }
+        }
+    }
+}
+
+impl Backend for ostree::Deployment {
+    fn backend<'a>(&self) -> Result<Box<dyn BackendImpl>> {
+        if let Some(origin) = self.origin().as_ref() {
+            origin.backend()
+        } else {
+            Err(anyhow!("Deployment has no origin"))
+        }
+    }
+}
+
+impl Backend for &glib::KeyFile {
```

Comment:
This part feels weird to me...why the indirection? Why not just inline this in the `impl Backend for ostree::Deployment`?